Bug: `store: unable to create user session: pq`

When accessing a miniflux instance running on my local network through safari on MacOS 11.2, I get the following error after login:

```
store: unable to create user session: pq: invalid input syntax for type inet: "fe80::14c2:f039:edc7:****%eth0"
```

In logs:
```
2021-02-07 14:29:10.465 GMT [214] ERROR:  invalid input syntax for type inet: "fe80::14c2:f039:edc7:****%eth0"
2021-02-07 14:29:10.465 GMT [214] STATEMENT:  INSERT INTO user_sessions (token, user_id, user_agent, ip) VALUES ($1, $2, $3, $4)
[ERROR] [HTTP:Internal Server Error] /login => store: unable to create user session: pq: invalid input syntax for type inet: "fe80::14c2:f039:edc7:effa%eth0"
```

Accessing miniflux directly through its local ip `192.168.*.*` works fine however. So I suspect the `.local` domain resolves to an ipv6 address which miniflux's backend doesn't allow for type `inet`.

Opening miniflux on the same machine  (through a .local domain) using Chrome works fine too.

I'm running PostgreSQL 13.1 and Miniflux 2.0.27.

--- a/http/request/client_ip.go
+++ b/http/request/client_ip.go
@@ -10,6 +10,14 @@ import (
 	"strings"
 )
 
+func dropIPv6zone(address string) string {
+	i := strings.IndexByte(address, '%')
+	if i != -1 {
+		address = address[:i]
+	}
+	return address
+}
+
 // FindClientIP returns client real IP address.
 func FindClientIP(r *http.Request) string {
 	headers := []string{"X-Forwarded-For", "X-Real-Ip"}
@@ -19,6 +27,7 @@ func FindClientIP(r *http.Request) string {
 		if value != "" {
 			addresses := strings.Split(value, ",")
 			address := strings.TrimSpace(addresses[0])
+			address = dropIPv6zone(address)
 
 			if net.ParseIP(address) != nil {
 				return address
@@ -27,12 +36,11 @@ func FindClientIP(r *http.Request) string {
 	}
 
 	// Fallback to TCP/IP source IP address.
-	var remoteIP string
-	if strings.ContainsRune(r.RemoteAddr, ':') {
-		remoteIP, _, _ = net.SplitHostPort(r.RemoteAddr)
-	} else {
+	remoteIP, _, err := net.SplitHostPort(r.RemoteAddr)
+	if err != nil {
 		remoteIP = r.RemoteAddr
 	}
+	remoteIP = dropIPv6zone(remoteIP)
 
 	// When listening on a Unix socket, RemoteAddr is empty.
 	if remoteIP == "" {

--- a/http/request/client_ip_test.go
+++ b/http/request/client_ip_test.go
@@ -19,6 +19,21 @@ func TestFindClientIPWithoutHeaders(t *testing.T) {
 	if ip := FindClientIP(r); ip != "192.168.0.1" {
 		t.Fatalf(`Unexpected result, got: %q`, ip)
 	}
+
+	r = &http.Request{RemoteAddr: "fe80::14c2:f039:edc7:edc7"}
+	if ip := FindClientIP(r); ip != "fe80::14c2:f039:edc7:edc7" {
+		t.Fatalf(`Unexpected result, got: %q`, ip)
+	}
+
+	r = &http.Request{RemoteAddr: "fe80::14c2:f039:edc7:edc7%eth0"}
+	if ip := FindClientIP(r); ip != "fe80::14c2:f039:edc7:edc7" {
+		t.Fatalf(`Unexpected result, got: %q`, ip)
+	}
+
+	r = &http.Request{RemoteAddr: "[fe80::14c2:f039:edc7:edc7%eth0]:4242"}
+	if ip := FindClientIP(r); ip != "fe80::14c2:f039:edc7:edc7" {
+		t.Fatalf(`Unexpected result, got: %q`, ip)
+	}
 }
 
 func TestFindClientIPWithXFFHeader(t *testing.T) {
@@ -37,6 +52,15 @@ func TestFindClientIPWithXFFHeader(t *testing.T) {
 	r = &http.Request{RemoteAddr: "192.168.0.1:4242", Header: headers}
 
 	if ip := FindClientIP(r); ip != "2001:db8:85a3:8d3:1319:8a2e:370:7348" {
+		t.Fatalf(`Unexpected result, got: %q`, ip)
+	}
+
+	// Test with single IPv6 address with zone
+	headers = http.Header{}
+	headers.Set("X-Forwarded-For", "fe80::14c2:f039:edc7:edc7%eth0")
+	r = &http.Request{RemoteAddr: "192.168.0.1:4242", Header: headers}
+
+	if ip := FindClientIP(r); ip != "fe80::14c2:f039:edc7:edc7" {
 		t.Fatalf(`Unexpected result, got: %q`, ip)
 	}
 

